### PR TITLE
rsx: Surface cache improvements 2 [part 2 of 3]

### PR DESCRIPTION
**Pull request by kd-11**
- Further improves surface cache hierarchy and data tracking. Preserves data more strictly by ensuring everything generated on the GPU is visible to other surfaces.
- Implements Read Color Buffers and Read Depth Buffer for Vulkan. Fixes missing data when Cell and RSX work together to render a scene.
- Adds a workaround to prevent race conditions between Cell and RSX in the middle of a transfer operation. This is superceded by later works, but as they are still incomplete, this can serve as a stop-gap measure.

### CODE REVIEW

**Comment by Xertified** [comment]
Surface cache improvements 2: Electric Boogaloo

**Comment by MarioSonic2987** [comment]
Appears a weird layer in GTA IV since this PR.
<details>
 <summary>Images</summary>

image
image
</details>

[RenderDoc VK](https://drive.google.com/open?id=1z84ZsuzcT8bRwsi_cy98LeIQBoTfRulp)

File: rpcs3/Emu/RSX/Common/surface_utils.h [hunk 1]
```diff
@@ -286,27 +294,102 @@ namespace rsx
...
-				if (!tag.first)
-					break;
+				e.second = *reinterpret_cast<u64*>(vm::g_sudo_addr + e.first);
+			}
+		}
 
-				if (tag.second != *reinterpret_cast<u64*>(vm::g_sudo_addr + tag.first))
+		void shuffle_tag()
+		{
+			memory_tag_samples[0].second = memory_tag_samples[0].second;
```
Comment (gamerforEA) [inline rpcs3/Emu/RSX/Common/surface_utils.h]: Self-assignment looks like a typo

Comment (kd-11) [reply rpcs3/Emu/RSX/Common/surface_utils.h]: Yea, missing ~

**Pull request merged**